### ISSUE TITLE
Migration from Raft storage fails with fsm timeout

### ISSUE THREAD

**Issue by TheCatFelix** [hashicorp/vault#8317]
**Describe the bug**
Using Vault 1.3.2 in the official Vault docker image, storage migration from Raft -> File fails with the following message: 

> Error migrating: error mounting 'storage_source': failed to create fsm: timeout

I have tried a few other combinations such as File -> Raft with the same behavior for the Raft backend. I can provide the logs as well if helpful. I did read #7198 which matches the symptoms exactly, however, the corresponding PR indicates the issue is fixed in 1.3.2

**To Reproduce**
Please find the detailed logs and script to repro attached at the bottom of this bug report.

**Expected behavior**
Storage backend is successfully migrated from Raft to File.

**Environment:**
Official Docker: Vault image 1.3.2
Debian Unstable with:
- Salt 2018.3.4
- Docker 18.09.1

Vault server configuration file(s):
Please find the configuration files in the detailed attachment.

**Additional context**
The script repro.sh first dumps all the files (vault configuration templates, salt state files) and then proceeds to setup a 3 node cluster with Raft storage backend from scratch. Node 1 (vault_1) is the leader of the cluster, nodes vault_2 and vault_3 are the standbys. After that, it attempts to migrate the Raft storage to File storage backend using vault_1.

```
root@debian-8:~/vault# ./repro.sh
#################### Show the code ####################
==> repro.sh <==
#!/bin/bash

echo '#################### Show the code ###################'#

head -n-0 *

echo '#################### Clean up ####################'

docker container rm -f vault_{1..3}
rm -rf /opt/files

echo '#################### Create cluster members ####################'

salt-call --local --file-root=. state.apply vault --state-output=terse 2> /dev/null

echo '#################### Setup the cluster ####################'

R="$(curl -sS -X PUT -d '{"secret_shares":1,"secret_threshold":1}' http://127.0.0.1:8201/v1/sys/init)"

export KEY=$(jq -r '.keys_base64 | .[]' <<< "$R")
export VAULT_TOKEN=$(jq -r '.root_token ' <<< "$R")

echo "Master Key -> $KEY"
echo "Root Token -> $VAULT_TOKEN"

curl -sS -X PUT -d '{"key":"'$KEY'"}' http://127.0.0.1:8201/v1/sys/unseal | jq

API_ADDR=$(salt-call --local grains.get ip4_interfaces --out=json 2> /dev/null | jq -r '.local | .docker0 | .[]')

sleep 10
curl -sS -X PUT -d '{"leader_api_addr":"http://'$API_ADDR':8201"}' http://127.0.0.1:820{2,3}/v1/sys/storage/raft/join | jq

sleep 3
curl -sS -X PUT -d '{"key":"'$KEY'"}' http://127.0.0.1:820{2,3}/v1/sys/unseal | jq

sleep 3
curl -sS -X GET -H "X-Vault-Token: $VAULT_TOKEN" http://127.0.0.1:8201/v1/sys/storage/raft/configuration | jq
curl -sS -X GET http://127.0.0.1:820{1,2,3}/v1/sys/health | jq

echo '#################### Attempt Backup ####################'

salt-call --local --file-root=. state.apply vault_migrate --state-output=mixed 2> /dev/null

==> vault.hcl.jinja <==
  ui = "true"
  storage "raft" {
    path    = "/vault/vault_{{ instance }}/"
    node_id = "vault_{{ instance }}"
  }
  listener "tcp" {
    address = ":8200"
    cluster_address = "0.0.0.0:8201"
    tls_disable = true
  }
  cluster_addr = "http://{{ grains.ip4_interfaces['docker0'] | first }}:821{{ instance }}"
  api_addr = "http://{{ grains.ip4_interfaces['docker0'] | first }}:820{{ instance }}"

==> vault_migrate_raft_to_file.hcl.jinja <==
  storage_source "raft" {
    path    = "/vault/vault_{{ instance }}/"
    node_id = "vault_{{ instance }}"
  }
  storage_destination "file" {
    path = "/backup_bind"
  }

==> vault_migrate.sls <==

{% set i = '1' %}

/opt/files/vault_backup:
 file.directory:
 - mode: 777

/opt/files/vault_backup_raft_to_file.hcl:
 file.managed:
 - source: salt://vault_migrate_raft_to_file.hcl.jinja
 - template: jinja
 - context:
   instance: {{ i }}

vault_migrate_raft_to_file:
 docker_container.run:
 - image: vault:1.3.2
 - command: operator migrate -config=/vault/config/vault_backup.hcl
 - replace: True
 - binds:
   - /opt/files/vault_backup:/backup_bind:rw
   - /opt/files/vault_backup_raft_to_file.hcl:/vault/config/vault_backup.hcl:ro
   - /opt/files/vault_{{ i }}:/vault/vault_{{ i }}:rw
 - cap_add: IPC_LOCK
 - require:
   - file: /opt/files/vault_backup_raft_to_file.hcl
   - file: /opt/files/vault_backup

==> vault.sls <==

{% set instances = [ '1', '2', '3' ] %}

/opt/files:
 file.directory

{% for i in instances %}

/opt/files/vault_{{ i }}:
 file.directory:
 - mode: 777

/opt/files/vault_{{ i }}.hcl:
 file.managed:
 - source: salt://vault.hcl.jinja
 - template: jinja
 - context:
   instance: {{ i }}

vault_{{ i }}:
 docker_container.running:
 - image: vault:1.3.2
 - command: server
 - port_bindings:
   - 820{{ i }}:8200
   - 821{{ i }}:8201
 - binds:
   - /opt/files/vault_{{ i }}:/vault/vault_{{ i }}:rw
   - /opt/files/vault_{{ i }}.hcl:/vault/config/vault.hcl:ro
 - cap_add: IPC_LOCK
 - restart_policy: unless-stopped
 - require:
   - file: /opt/files/vault_{{ i }}.hcl

{% endfor %}
#################### Clean up ####################
vault_1
vault_2
vault_3
#################### Create cluster members ####################
local:
  Name: /opt/files - Function: file.directory - Result: Changed Started: - 12:00:45.515628 Duration: 3.098 ms
  Name: /opt/files/vault_1 - Function: file.directory - Result: Changed Started: - 12:00:45.518805 Duration: 0.75 ms
  Name: /opt/files/vault_1.hcl - Function: file.managed - Result: Changed Started: - 12:00:45.519622 Duration: 22.867 ms
  Name: vault_1 - Function: docker_container.running - Result: Changed Started: - 12:00:45.566884 Duration: 747.151 ms
  Name: /opt/files/vault_2 - Function: file.directory - Result: Changed Started: - 12:00:46.314235 Duration: 1.094 ms
  Name: /opt/files/vault_2.hcl - Function: file.managed - Result: Changed Started: - 12:00:46.315412 Duration: 26.632 ms
  Name: vault_2 - Function: docker_container.running - Result: Changed Started: - 12:00:46.342537 Duration: 848.769 ms
  Name: /opt/files/vault_3 - Function: file.directory - Result: Changed Started: - 12:00:47.192257 Duration: 4.179 ms
  Name: /opt/files/vault_3.hcl - Function: file.managed - Result: Changed Started: - 12:00:47.197003 Duration: 23.734 ms
  Name: vault_3 - Function: docker_container.running - Result: Changed Started: - 12:00:47.221097 Duration: 801.761 ms

Summary for local
-------------
Succeeded: 10 (changed=10)
Failed:     0
-------------
Total states run:     10
Total run time:    2.480 s
#################### Setup the cluster ####################
Master Key -> oyavqjghSPQsR2EXIjkfFrAEGBo4oJhCVHX0euMS4sQ=
Root Token -> s.Sw84J4QKoF9M8H6b4K7Xk5li
{
  "type": "shamir",
  "initialized": true,
  "sealed": false,
  "t": 1,
  "n": 1,
  "progress": 0,
  "nonce": "",
  "version": "1.3.2",
  "migration": false,
  "cluster_name": "vault-cluster-8d0e9838",
  "cluster_id": "241bda99-9208-1283-538b-2ccee5242971",
  "recovery_seal": false,
  "storage_type": "raft"
}
{
  "joined": true
}
{
  "joined": true
}
{
  "type": "shamir",
  "initialized": true,
  "sealed": true,
  "t": 1,
  "n": 1,
  "progress": 0,
  "nonce": "",
  "version": "1.3.2",
  "migration": false,
  "recovery_seal": false,
  "storage_type": "raft"
}
{
  "type": "shamir",
  "initialized": true,
  "sealed": true,
  "t": 1,
  "n": 1,
  "progress": 0,
  "nonce": "",
  "version": "1.3.2",
  "migration": false,
  "recovery_seal": false,
  "storage_type": "raft"
}
{
  "request_id": "b094db96-0726-6f92-e564-873c499b90e4",
  "lease_id": "",
  "renewable": false,
  "lease_duration": 0,
  "data": {
    "config": {
      "servers": [
        {
          "node_id": "vault_1",
          "address": "172.17.0.1:8211",
          "leader": true,
          "protocol_version": "3",
          "voter": true
        },
        {
          "node_id": "vault_2",
          "address": "172.17.0.1:8212",
          "leader": false,
          "protocol_version": "3",
          "voter": true
        },
        {
          "node_id": "vault_3",
          "address": "172.17.0.1:8213",
          "leader": false,
          "protocol_version": "3",
          "voter": true
        }
      ],
      "index": 29
    }
  },
  "wrap_info": null,
  "warnings": null,
  "auth": null
}
{
  "initialized": true,
  "sealed": false,
  "standby": false,
  "performance_standby": false,
  "replication_performance_mode": "disabled",
  "replication_dr_mode": "disabled",
  "server_time_utc": 1581105667,
  "version": "1.3.2",
  "cluster_name": "vault-cluster-8d0e9838",
  "cluster_id": "241bda99-9208-1283-538b-2ccee5242971"
}
{
  "initialized": true,
  "sealed": false,
  "standby": true,
  "performance_standby": false,
  "replication_performance_mode": "unknown",
  "replication_dr_mode": "unknown",
  "server_time_utc": 1581105667,
  "version": "1.3.2",
  "cluster_name": "vault-cluster-8d0e9838",
  "cluster_id": "241bda99-9208-1283-538b-2ccee5242971"
}
{
  "initialized": true,
  "sealed": false,
  "standby": true,
  "performance_standby": false,
  "replication_performance_mode": "unknown",
  "replication_dr_mode": "unknown",
  "server_time_utc": 1581105667,
  "version": "1.3.2",
  "cluster_name": "vault-cluster-8d0e9838",
  "cluster_id": "241bda99-9208-1283-538b-2ccee5242971"
}
#################### Attempt Backup ####################
local:
  Name: /opt/files/vault_backup - Function: file.directory - Result: Changed Started: - 12:01:09.744150 Duration: 3.948 ms
  Name: /opt/files/vault_backup_raft_to_file.hcl - Function: file.managed - Result: Changed Started: - 12:01:09.748187 Duration: 22.248 ms
----------
          ID: vault_migrate_raft_to_file
    Function: docker_container.run
      Result: False
     Comment: Container ran and exited with a return code of 2
     Started: 12:01:09.795620
    Duration: 2982.706 ms
     Changes:
              ----------
              ExitCode:
                  2
              Id:
                  70d206c31a716de43aeba7b1d4e778c5e2bf816fdbb6aa26c508a999fef08e5f
              Logs:
                  Error migrating: error mounting 'storage_source': failed to create fsm: timeout
              Name:
                  vault_migrate_raft_to_file
              Replaces:
                  - be01907863e0a7da192e53577b2c6583cbffd6dfb543914f51f608cad438ef66
              Time_Elapsed:
                  1.831688404083252
              Warnings:
                  None

Summary for local
------------
Succeeded: 2 (changed=3)
Failed:    1
------------
Total states run:     3
Total run time:   3.009 s
root@debian-8:~/vault#
```

**Comment by briankassouf** [comment]
@TheCatFelix Thank you for the detailed report. 

Could you try shutting down the original nodes that you created before attempting the migration? The underlying data file uses file locks to ensure only one process is accessing the file. I believe the timeout will go away once the other process is terminated. 

Additionally i see the title "Attempt Backup" before doing the migration. If you're looking for a backup solution we allow snapshots to be take through vault's API. This API does not require the node to be shutdown. Docs for the snapshot APIs are here: https://www.vaultproject.io/api-docs/system/storage/raft/#take-a-snapshot-of-the-raft-cluster. Also the following CLI command:
```
vault operator raft snapshot save backup.snap
```

**Comment by TheCatFelix** [comment]
@briankassouf Thanks for taking a look quickly!

More than happy to retry, unfortunately results have been negative too:

`Error migrating: error setting migration lock: raft storage backend is not initialized`

Details of the execution here:

```#################### Attempt Backup ####################
local:
  Name: vault_3 - Function: docker_container.stopped - Result: Changed Started: - 19:15:57.871273 Duration: 622.354 ms
  Name: vault_2 - Function: docker_container.stopped - Result: Changed Started: - 19:15:58.493806 Duration: 472.3 ms
  Name: vault_1 - Function: docker_container.stopped - Result: Changed Started: - 19:15:58.966293 Duration: 2956.652 ms
  Name: /opt/files/vault_backup - Function: file.directory - Result: Changed Started: - 19:16:01.928201 Duration: 9.328 ms
  Name: /opt/files/vault_backup_raft_to_file.hcl - Function: file.managed - Result: Changed Started: - 19:16:01.937866 Duration: 33.503 ms
----------
          ID: vault_migrate_raft_to_file
    Function: docker_container.run
      Result: False
     Comment: Container ran and exited with a return code of 2
     Started: 19:16:01.971960
    Duration: 2352.752 ms
     Changes:
              ----------
              ExitCode:
                  2
              Id:
                  58897b9349a465fe5bb2b338b45a5b66408c43bdd406f5cde0c5e4776e27eec0
              Logs:
                  Error migrating: error setting migration lock: raft storage backend is not initialized
              Name:
                  vault_migrate_raft_to_file
              Replaces:
                  - d8d03c2bb359e3c8e768d903a0ab2b1cf0779e42769b851b872368f0ca3a9c68
              Time_Elapsed:
                  0.8969099521636963
              Warnings:
                  None

Summary for local
------------
Succeeded: 5 (changed=6)
Failed:    1
------------
Total states run:     6
Total run time:   6.447 s
```

To shut down the containers before the migration, I just had SALT stop them gracefully as follows in the `vault_migrate.sls`. For good measure, making sure the master gets shut down last. I had also tried inserting a 10s sleep between container shut down and migration, same result:

```
{% for i in [ '3', '2', '1' ] %}

vault_{{ i }}:
 docker_container.stopped:
 - require_in:
   - docker_container: vault_migrate_raft_to_file

{% endfor %}
```

Just to make sure that there is actually data in /opt/files/vault_1:

```
root@debian-8:~/vault# find /opt/files/vault_1
/opt/files/vault_1
/opt/files/vault_1/raft
/opt/files/vault_1/raft/snapshots
/opt/files/vault_1/raft/raft.db
/opt/files/vault_1/vault.db
```

As for the RAFT snapshot feature, I am aware of it and I am looking to switch to it once we have some mileage with RAFT backend in production.
Today we use the file backend, and have automation in place that can recover our primary vault in case of bigger disasters very easily. Vault being so central to everything, I would like to test RAFT and still have the possibility to save backups in our old format which makes this more of a "two-way-door" - easy to go back if something goes wrong.

**Comment by briankassouf** [comment]
Should be fixed by #8485. Going to close for now, but please do let us know if this isn't fixed for you

**Comment by TheCatFelix** [comment]
It appears fixed in master -  thank you very much.

Validated in the following way:

- Downloaded current master tip of tree
```
test@debian-8:~/vlt/vault$ git rev-parse HEAD
351d3ace7dece526efdcc85b98a091a976325667
```
- make dev
```
test@debian-8:~/vlt/vault$ /home/test/go/bin/vault -v
Vault v1.4.0-beta1 ('351d3ace7dece526efdcc85b98a091a976325667')
```
- "Hot patched" the containers with the new binary:
```
vault_{{ i }}:
 docker_container.running:
 - image: vault:1.3.2
 - command: server
 - port_bindings:
   - 820{{ i }}:8200
   - 821{{ i }}:8201
 - binds:
   - /home/test/go/bin/vault:/bin/vault:rw <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
   - /opt/files/vault_{{ i }}:/vault/vault_{{ i }}:rw
   - /opt/files/vault_{{ i }}.hcl:/vault/config/vault.hcl:ro
 - cap_add: IPC_LOCK
 - restart_policy: unless-stopped
 - require:
   - file: /opt/files/vault_{{ i }}.hcl
```

Now it does seem to run the container just fine to take the raft snapshot:
```
----------
          ID: vault_migrate_raft_to_file
    Function: docker_container.run
      Result: True
     Comment: Container ran and exited with a return code of 0
     Started: 12:53:21.551512
    Duration: 3696.544 ms
     Changes:
              ----------
              ExitCode:
                  0
              Id:
                  67730c3598bca37705c9537285a33639f2edeb6bbab33598d640ef61c012ca67
              Logs:
                  2020-03-12T19:53:24.543Z [INFO]  copied key: path=core/audit
                  2020-03-12T19:53:24.543Z [INFO]  copied key: path=core/auth
                  2020-03-12T19:53:24.544Z [INFO]  copied key: path=core/cluster/feature-flags
                  2020-03-12T19:53:24.545Z [INFO]  copied key: path=core/cluster/local/info
                  2020-03-12T19:53:24.545Z [INFO]  copied key: path=core/hsm/barrier-unseal-keys
                  2020-03-12T19:53:24.546Z [INFO]  copied key: path=core/keyring
                  2020-03-12T19:53:24.546Z [INFO]  copied key: path=core/leader/40b0288c-b097-3db3-9699-2a882efac43e
                  2020-03-12T19:53:24.547Z [INFO]  copied key: path=core/local-audit
                  2020-03-12T19:53:24.548Z [INFO]  copied key: path=core/local-auth
                  2020-03-12T19:53:24.548Z [INFO]  copied key: path=core/local-mounts
                  2020-03-12T19:53:24.549Z [INFO]  copied key: path=core/master
                  2020-03-12T19:53:24.551Z [INFO]  copied key: path=core/mounts
                  2020-03-12T19:53:24.552Z [INFO]  copied key: path=core/raft/tls
                  2020-03-12T19:53:24.553Z [INFO]  copied key: path=core/seal-config
                  2020-03-12T19:53:24.554Z [INFO]  copied key: path=core/shamir-kek
                  2020-03-12T19:53:24.554Z [INFO]  copied key: path=core/wrapping/jwtkey
                  2020-03-12T19:53:24.555Z [INFO]  copied key: path=logical/606edd55-448f-8b56-2322-573a9bd82e1e/casesensitivity
                  2020-03-12T19:53:24.555Z [INFO]  copied key: path=sys/policy/control-group
                  2020-03-12T19:53:24.555Z [INFO]  copied key: path=sys/policy/default
                  2020-03-12T19:53:24.556Z [INFO]  copied key: path=sys/policy/response-wrapping
                  2020-03-12T19:53:24.556Z [INFO]  copied key: path=sys/token/accessor/11a0368b7b9926608b51d221b790e0d0d32ab45f
                  2020-03-12T19:53:24.556Z [INFO]  copied key: path=sys/token/id/h2b1a779cf9fc55e606d69bc7c05aad487e7a765a30eb061dca2789d4d37b108e
                  2020-03-12T19:53:24.557Z [INFO]  copied key: path=sys/token/salt
                  Success! All of the keys have been migrated.
              Name:
                  vault_migrate_raft_to_file
              Replaces:
                  - fc5c1d6c23d62c7bb99ec9c04591b05e573fb9c759b7208903e7734fad8f74ac
              Time_Elapsed:
                  1.4356088638305664
              Warnings:
                  None

Summary for local
------------
Succeeded: 6 (changed=1)
Failed:    0
------------
Total states run:     6
Total run time:   3.795 s
```
